### PR TITLE
Add read/write methods for date, time, and datetime types

### PR DESCRIPTION
**Pull request by daviditen**
Add `readWriteThis` methods for the date, time, datetime types. Add a test to
ensure that the methods can be called and that they output and input matching
values when appropriate.

### CODE REVIEW

**Comment by buddha314** [comment]
Go David Go!!

**Comment by marcoscleison** [comment]
This issue is important. Thanks!

File: modules/standard/DateTime.chpl [hunk 1]
```diff
@@ -434,6 +434,13 @@ module DateTime {
     return str;
   }
 
+  /* Read or write a date value from channel `f` */
+  proc date.readWriteThis(f) {
+    var dash = new ioLiteral("-");
+    f <~> new ioLiteral("{") <~> chpl_year <~> dash <~> chpl_month <~> dash
+      <~> chpl_day <~> new ioLiteral("}");
```
Comment (mppf) [inline modules/standard/DateTime.chpl]: Is there a specific reason for the { } around the date? I'd think 2018-03-12 is clear enough as a date already. Are you following some Python precedent, e.g.?

File: modules/standard/DateTime.chpl [hunk 2]
```diff
@@ -658,6 +664,15 @@ module DateTime {
     return str;
   }
 
+  /* Read or write a time value from channel `f` */
+  proc time.readWriteThis(f) {
+    var colon = new ioLiteral(":");
+    f <~> new ioLiteral("{") <~> chpl_hour <~> colon <~> chpl_minute <~> colon
+      <~> chpl_second <~> new ioLiteral(".") <~> chpl_microsecond
+      <~> new ioLiteral("}");
```
Comment (mppf) [inline modules/standard/DateTime.chpl]: Same question here about {.

File: test/library/standard/DateTime/readWriteDatesTimes.good [hunk 3]
```diff
@@ -0,0 +1,3 @@
+datetime: true
+date: true
+time: true
```
Comment (mppf) [inline test/library/standard/DateTime/readWriteDatesTimes.good]: Do we have any tests that actually show the date-time formats in the .good file? Might be worth adding one so we can be sure it doesn't accidentally change.

**Comment by mppf** [comment]
Should this be following [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601)?

**Comment by buddha314** [comment]
@mppf you probably have a point.  I'm not sure the proper way to set it.

**Comment by mppf** [comment]
The readWriteThis methods can check for json like so:
``` chapel
      var binary = f.binary();
      var arrayStyle = f.styleElement(QIO_STYLE_ELEMENT_ARRAY);
      var isjson = arrayStyle == QIO_ARRAY_FORMAT_JSON && !binary;
```

File: modules/standard/DateTime.chpl (hunk 1 above)
Comment (daviditen) [reply modules/standard/DateTime.chpl]: I was following the default behavior of Chapel classes, which doesn't make a whole lot of sense given that these are records.  I'll remove the curly braces.

File: test/library/standard/DateTime/readWriteDatesTimes.good (hunk 3 above)
Comment (daviditen) [reply test/library/standard/DateTime/readWriteDatesTimes.good]: I'll add one.

**Comment by daviditen** [comment]
> Should this be following ISO 8601?

This means zero-padding the year to 4 digits and everything else to 2 digits, plus putting a 'T' between the date and time, right?  The padding sounds fine, but I'm not crazy about the 'T'.

**Comment by mppf** [comment]
> > Should this be following ISO 8601?
> 
> This means zero-padding the year to 4 digits and everything else to 2 digits, plus putting a 'T' between the date and time, right?  The padding sounds fine, but I'm not crazy about the 'T'.

I think so... and generally agree. But at least ISO 8601 is a standard.

**Comment by mppf** [comment]
Looking good, thanks.

**Comment by daviditen** [comment]
Thank you @mppf!

**Pull request merged**